### ISSUE TITLE
Namespace/module documentation

### ISSUE THREAD

**Issue by rendaw** [caddyserver/caddy#4242]
Using the instructions for automatic https with a reverse proxy, I get warnings like `missing email address for ZeroSSL; it is strongly recommended to set one for next time` in the logs.  AFAICT it seems like it's probably a good idea to provide an email for automatic https cert provisioning.

I found https://caddyserver.com/docs/modules/tls.issuance.acme after googling, but it's not clear where in the configuration this goes.  I'm using JSON config luckily, I don't see anything about using this with non-JSON config. Putting it in ".tls.issuance.acme" using jq syntax resulted in the error `json: unknown field \"tls\"","channel`.

After lots of digging I found https://caddyserver.com/docs/json/apps/tls/automation/policies/issuers/ which says `Fulfilled by modules in namespace: tls.issuance` which seems to mean this is the place that config goes, however clicking on the links to the `acme` sub-object links to a different page, https://caddyserver.com/docs/json/apps/tls/automation/policies/issuers/acme/ .

To summarize:
1. Have multiple documentation pages for each module with the same information is confusing
2. There's no easy way to find the mapping between namespaces and where to put stuff in config if you don't already know where it goes
3. Can that warning be ignored?  If not, should this be considered best practice and documented somewhere prominently (perhaps with the automatic provisioning examples elsewhere in the docs).  It's my understanding that at least LE would prefer people to provide contact info (which I think is this email).
4. I can open a separate issue for this, but if specifying an email is best practice, it seems like it requires a lot of extra boilerplate config.  Is there a way this could be simpler?

**Comment by francislavoie** [comment]
1. The JSON docs are auto-generated from the codebase, including from community-maintained plugins that are registered (see https://caddyserver.com/download for the list of plugins). The two different locations serve different purposes -- the direct module page is for https://caddyserver.com/docs/modules/ which is an index of all the modules that exist, which is handy for Ctrl+F search/reference. The https://caddyserver.com/docs/json/ page is a full tree/graph view of the JSON configuration, from the top-down.

2. Look at the path in the URL (or the breadcrumb at the top of that page). That's how you know where it goes. The config is a heavily nested structure, so I'm not sure how else it could be made clearer.

3. Yes, the warning can be ignored. Caddy's gotten permission from both Let's Encrypt and ZeroSSL for making the email field optional, since Caddy's well automated and will generally complain in the logs if something went wrong, and we want Caddy's UX to be as easy as possible. But yes, having an email configured is still very much recommended.

4. It's very simple to configure in the Caddyfile, via the `email` global option. There's not really any feasible way for making it simpler in the JSON config, because it's a deeply nested config option, that must be set for two different issuers. The Caddyfile config scripts applying it to both Let's Encrypt and ZeroSSL issuers. It's as simple as this:

```
{
	email you@example.com
}

example.com {
	respond "Hullo!" 200
}
```

You can run `caddy adapt --pretty` on this `Caddyfile` to see the JSON output:

```json
{
  "apps": {
    "http": {
      "servers": {
        "srv0": {
          "listen": [
            ":443"
          ],
          "routes": [
            {
              "match": [
                {
                  "host": [
                    "example.com"
                  ]
                }
              ],
              "handle": [
                {
                  "handler": "subroute",
                  "routes": [
                    {
                      "handle": [
                        {
                          "body": "Hullo!",
                          "handler": "static_response",
                          "status_code": 200
                        }
                      ]
                    }
                  ]
                }
              ],
              "terminal": true
            }
          ]
        }
      }
    },
    "tls": {
      "automation": {
        "policies": [
          {
            "subjects": [
              "example.com"
            ],
            "issuers": [
              {
                "email": "you@example.com",
                "module": "acme"
              },
              {
                "email": "you@example.com",
                "module": "zerossl"
              }
            ]
          }
        ]
      }
    }
  }
}
```

If you're having trouble writing JSON configuration, I recommend starting with the Caddyfile and using the `adapt` command to figure out what valid config it generates, then going from there as your base. Much easier to get started.

Also, you can use the https://github.com/abiosoft/caddy-json-schema tool which generates a JSON schema that you can import into your IDE (VS Code, etc) to give you automatic completion for writing JSON config (or the YAML or TOML adapters which use the same structure).

For next time, please ask your usage questions on the [Caddy community forums](https://caddy.community). We prefer to keep the GitHub issue board for bugs and feature requests. Don't forget to fill out the thread template so we can help you!

**Comment by rendaw** [comment]
Thanks for the response! This ticket was a request to make the documentation clearer, I don't agree that it was a question. 
It's clear that you put a lot of effort into the response - but everything you posted here would be immensely useful to have in the documentation.

I feel like perhaps the points I was making in 1 and 2 were lost, so please let me know if you'd like clarification.

**Comment by francislavoie** [comment]
Well, I didn't really see anything actionable in your issue. For documentation changes, we need to be talking about specifics. The docs can be found in https://github.com/caddyserver/website, feel free to propose any specific changes there via a pull request.

Re-reading your issue, I realized there might be a disconnect which I didn't properly explain between module namespaces and config paths. Those are somewhat unrelated things. I recommend you read through https://caddyserver.com/docs/architecture to get an idea of how it works.

**Comment by rendaw** [comment]
Awesome, and thanks for taking another look.  

> Re-reading your issue, I realized there might be a disconnect which I didn't properly explain between module namespaces and config paths. Those are somewhat unrelated things.

Exactly! I requested explicitly documenting the mapping from module namespaces to config paths because as you stated there's know way to know one from the other.   I read through https://caddyserver.com/docs/architecture as well as https://caddyserver.com/docs/extending-caddy/namespaces before filing this issue, neither of them indicate where to look in the config for where the module config goes.

My specific suggestions would be:
1. Instead of linking to `https://caddyserver.com/docs/modules/tls.issuance.acme`, the links in the modules page should link to ex: `https://caddyserver.com/docs/json/apps/tls/automation/policies/issuers/acme/`, because the former doesn't contain information on where the config goes in the breadcrumbs or url but the latter does, and the bodies of both pages are the same.  Perhaps the namespace could also be added by the generator.
2. The example you provided:
    ```
   {
       # You can leave this out, but we recommend setting it!
       email you@example.com
   }

   example.com {
       respond "Hello, privacy!" 200
   }
   ```
   should replace
   ```
   example.com

   respond "Hello, privacy!"
   ```
   on https://caddyserver.com/docs/quick-starts/https with a note about what it's for and the agreement with LE and ZeroSSL.

   On the other hand, if it's 100% okay not to set it the warning should be eliminated since it contradicts this indicating there's a non-fatal problem with not setting it.

**Comment by rendaw** [comment]
Just to be clear, this is an awesome project and I really appreciate everything you're doing here.  I wanted to leave feedback here because I've written a lot of documentation before and time and time again get asked simple questions, then go back and find I was missing something or was unclear in my explanations etc. and nobody ever mentioned it to me.

**Comment by mholt** [comment]
@rendaw 

> Instead of linking to https://caddyserver.com/docs/modules/tls.issuance.acme, the links in https://caddyserver.com/docs/modules/ should link to ex: https://caddyserver.com/docs/json/apps/tls/automation/policies/issuers/acme/, because the former doesn't contain information on where the config goes in the breadcrumbs or url but the latter does, and the bodies of both pages are the same. Perhaps the namespace could also be added by the generator.

The problem is that modules can appear in more than one place, and we don't know which is the right place for what the user is looking up. Only the user knows that. A module's documentation and where it goes in the config structure are unrelated, so it's important we link to the module docs rather than assuming a (wrong) spot in the structure.

I do agree it would be cool on the `/modules/` pages to show where in the config structure we are aware that module can be used; but I am not sure of an efficient way to implement that yet. It would take a lot of time to figure that out and implement it, but it's not a priority right now because of other projects.

**Comment by francislavoie** [comment]
> Instead of linking to https://caddyserver.com/docs/modules/tls.issuance.acme, the links in caddyserver.com/docs/modules should link to ex: https://caddyserver.com/docs/json/apps/tls/automation/policies/issuers/acme/, because the former doesn't contain information on where the config goes in the breadcrumbs or url but the latter does, and the bodies of both pages are the same. Perhaps the namespace could also be added by the generator.

That's non-trivial because the module may be used in multiple different places (sometimes recursively, such as `http.handlers` modules since `http.handlers.subroute` can contain other handlers).

> The example you provided [...] should replace [...] on caddyserver.com/docs/quick-starts/https with a note about what it's for and the agreement with LE and ZeroSSL.

I disagree. The purpose of the quick start docs is to show the simplest way to get up and running. Including optional config is orthogonal to this idea.

> On the other hand, if it's 100% okay not to set it the warning should be eliminated since it contradicts this indicating there's a non-fatal problem with not setting it.

A warning is a warning. A warning is not an error, it's not fatal. There's no harm in ignoring warnings.

**Comment by mholt** [comment]
> There's no harm in ignoring warnings.

Well, there _could_ be later. That's kind of the idea. If something goes seriously wrong leading to possible disruption of service, and the CA needs to get a hold of you, they won't be able to without an email address.

**Comment by rendaw** [comment]
Thanks for the responses and explaining the difficulties of these things.  This all sounds reasonable to me, although it's unfortunate how things can come up in search results etc. but I really appreciate that you took time to read this all through.
